### PR TITLE
Bundle scroll-time polyfill into library

### LINKED ISSUE #572: Content Security Policy violation for script-src
#### What are you trying to do?

I'm using ngx-scrollbar for my web-extenstion pet-project. It works as intended until I try to actually build and pack it into browser extension:

Снимок экрана 2024-04-08 в 07 39 08

As you could see whenever it's loaded as extension I receive this CSP warning with message about failing to load remote script which refers to `https://flackr.github.io/scroll-timeline/dist/scroll-timeline.js`.

#### What troubleshooting steps have you tried?

It fixes with no issues if I update my extension's `manifest.json` CSP rules with something like:

```json
    "content_security_policy": "script-src 'self' https://flackr.github.io"
```

...but in fact I don't really want to do it, because remote code is severely dangerous thing to leave. And it will cause review issues if I actually try to release it to stores like AMO or CWS.

Could we somehow pack this code into ngx-scrollbar package please?

#### Reproduction

Stackblitz example won't really help here because you couldn't reproduce this issue there (you need this code to be loaded as web extension). But if you need it I would create reproduction repository with instructions

#### Environment

- Angular: 17
- ngx-scrollbar: 14
- Browser(s): Firefox 124
- Operating System (e.g. Windows, macOS, Ubuntu): any

### PR DESCRIPTION
**Pull request by Smarthard**
resolves #572

I was able to figure out why tests were failing. Karma just wasn't serving new library's assets directory. Sorry to kept you waiting.

I added polyfill as is. Not sure if we actually need to provide a license notice or something like that there. But as soon as original repository will provide a proper npm package we could be able to remove it from here and work with it as a plain dependency.

### CODE REVIEW

**Comment by MurhafSousli** [comment]
@Smarthard Thanks for the PR! could you double check why local coverage is broken

**Comment by Smarthard** [comment]
@MurhafSousli not sure. Probably your CI workflow missing `write` permission. At least that's what I found at [marocchino/sticky-pull-request-comment readme](https://github.com/marocchino/sticky-pull-request-comment?tab=readme-ov-file#basic). Here is the related issue: https://github.com/marocchino/sticky-pull-request-comment/issues/298

**Comment by MurhafSousli** [comment]
@Smarthard I opened a PR to test the build, it has no issue https://github.com/MurhafSousli/ngx-scrollbar/pull/585

Weird, perhaps it couldn't comment on the PR cause it is opened by you?

**Comment by Smarthard** [comment]
I created a test branch with changes similar to yours in #585 and opened PRs #586 here and https://github.com/Smarthard/ngx-scrollbar/pull/1 in my fork to test if it really a permission's issue

**Comment by Smarthard** [comment]
@MurhafSousli Yeah looks like you were right. My changes passed workflow in my repo and failed in yours. I think you need something like this to solve this issues with permissions.

```patch
diff --git a/.github/workflows/integrate.yml b/.github/workflows/integrate.yml
index 7ffcc0e..2729b37 100644
--- a/.github/workflows/integrate.yml
+++ b/.github/workflows/integrate.yml
@@ -8,6 +8,10 @@ on:
 
 jobs:
   build:
+    # Allow forks to be annotated with code coverage properly
+    permissions:
+      pull-requests: write
+
     # Machine environment:
     # We specify the Node.js version manually below, and use versioned Chrome from Puppeteer.
     runs-on: ubuntu-latest
```

**Comment by MurhafSousli** [comment]
I fixed the CI build issue, I got a question, the script is copied to assets directory of the plugin. but how would be copied to the user's app assets directory? would that happen automatically?

**Comment by Smarthard** [comment]
> I fixed the CI build issue, I got a question, the script is copied to assets directory of the plugin. but how would be copied to the user's app assets directory? would that happen automatically?

By including it from `angular.json` build options as any other assets I suppose. Yeah it's a breaking change

**Comment by MurhafSousli** [comment]
> > I fixed the CI build issue, I got a question, the script is copied to assets directory of the plugin. but how would be copied to the user's app assets directory? would that happen automatically?
> 
> By including it to `angular.json` build options as any other assets I suppose. Yeah it's a breaking change
> 
> UPD something like this should work (I removed unnecessary lines):
> 
> ```json
> {
>     "projects": {
>         "app": {
>             "architect": {
>                 "build": {
>                     "builder": "@angular-devkit/build-angular:browser",
>                     "options": {
>                         "assets": [
>                             {
>                                 "glob": "**/*",
>                                 "input": "node_modules/ngx-scrollbar/assets/",
>                                 "output": "assets"
>                             }
>                         ],
>                     }
>                 }
>             }
>         }
>     }
> }
> ```

In this case, I want to keep loading the script from the github link as the default option, without breaking changes and the only change in this PR would be the shipping of the script with the package.

We can add to the documentation the changes needed to load it manually from the assets directory.

File: projects/ngx-scrollbar/karma.conf.js
Comment (MurhafSousli) [inline projects/ngx-scrollbar/karma.conf.js:1]: Reset

File: projects/ngx-scrollbar/src/lib/tests/scrollbar-manager.spec.ts
Comment (MurhafSousli) [inline projects/ngx-scrollbar/src/lib/tests/scrollbar-manager.spec.ts:1]: Reset

File: projects/ngx-scrollbar/src/lib/utils/scrollbar-manager.ts
Comment (MurhafSousli) [inline projects/ngx-scrollbar/src/lib/utils/scrollbar-manager.ts:1]: Reset

File: projects/ngx-scrollbar-demo/server.ts
Comment (MurhafSousli) [inline projects/ngx-scrollbar-demo/server.ts:1]: Reset

**Review by MurhafSousli — CHANGES_REQUESTED** [review]
In this case, I want to keep loading the script from the github link as the default option, without breaking changes and the only change in this PR would be the shipping of the script with the package.

We can add to the documentation the changes needed to load it manually from the assets directory.

**Comment by Smarthard** [comment]
@MurhafSousli Hello!

I updated PR as you required. All changes were squashed into single commit and force-pushed to make it clearer.

**Comment by MurhafSousli** [comment]
@leblancmeneses I don't get it, in Chrome the polypill script will not even load, and the URL is stored as a string value. what is the complaint about?

**Comment by MurhafSousli** [comment]
@Smarthard Sorry for the late response. I added the documentation here https://github.com/MurhafSousli/ngx-scrollbar/wiki/Scroll-timeline-polyfill

**Comment by Smarthard** [comment]
@MurhafSousli no worries! Thanks! I just tested and it's working great.

One good thing to point out for anyone who will use it that way. I needed to additionally modify my CSP rules for style-src like this: `style-src 'unsafe-inline' blob:`. That's because polyfill trying to dynamically assign styles via `createObjectURL`.

For future references I'd like to left an example how I handled these changes for my project: https://github.com/Smarthard/shikicinema/commit/c5ae774c92782718078c1b407a007912730f3a49

**Comment by Smarthard** [comment]
@leblancmeneses frankly, from my experience moderators from Chrome Web Store often falsy-positive report issues on items to their developers. My recommendation is to reply to CWS from email message you receive (look closely for reply contacts). You could describe there your concerns about false-positive alert and provide them references from your code and/or references from the software your are using.

**Comment by MurhafSousli** [comment]
@Smarthard  Is this related https://github.com/MurhafSousli/ngx-scrollbar/issues/590 or this https://github.com/MurhafSousli/ngx-scrollbar/issues/594 related?

**Comment by Smarthard** [comment]
> Is https://github.com/MurhafSousli/ngx-scrollbar/issues/590 or https://github.com/MurhafSousli/ngx-scrollbar/issues/594 related?

I think I've seen those kind of errors in my console as well, but after CSP correction there are no errors.

>It would be beneficial if you could add a how-to make it work with CSP at the end at this documentation

For common SSR/CSR angular web apps this should be sufficient:

```html
<!-- index.html -->
<meta http-equiv="Content-Security-Policy" content="style-src 'unsafe-src' blob:">
```

For web extensions you would need to modify `manifest.json` of your extension, changes to `index.html` are not necessary in this case:

```json
"content_security_policy": "style-src 'unsafe-src' blob:"
```

**Comment by leblancmeneses** [comment]
@MurhafSousli

After removing the string with patch-package, my Chrome extension was approved. 
https://chromewebstore.google.com/detail/tabsift/gihejcnfnbdiohnjepkkdhilfedjcahm?hl=en&authuser=0

I think the best solution here would be to bump the major version and change NG_SCROLLBAR_POLYFILL

```
export type PolyFillLoader = (url: string) => Promise<void>;
export const NG_SCROLLBAR_POLYFILL = new InjectionToken<PolyFillLoader>('NG_SCROLLBAR_POLYFILL');
```

and figure out a way to make it tree-shakable for consumers wanting to override the current default behavior.

Thanks for maintaining this!

@Smarthard , This violation is pretty clear 'modules pointing to external files'.

**Comment by MurhafSousli** [comment]
So if any variable contains a string value of an external URL would be rejected even if it is never used at runtime! that's weird

**Pull request merged**